prow/userdashboard/userdashboard.go

```diff
@@ -0,0 +1,171 @@
+type DashboardAgent struct {
+	goac *config.GithubOAuthConfig
+
+	log *logrus.Entry
+}
+
+type Label struct {
+	ID   githubql.ID
+	Name githubql.String
+}
+
+type PullRequest struct {
+	Number githubql.Int
+	Author struct {
+		Login githubql.String
+	}
+	Repository struct {
+		Name          githubql.String
+		NameWithOwner githubql.String
+		Owner         struct {
+			Login githubql.String
+		}
+	}
+	Labels struct {
+		Nodes []struct {
+			Label Label `graphql:"... on Label"`
+		}
+	} `graphql:"labels(first: 100)"`
+}
+
+type PullRequestQuery struct {
+	Viewer struct {
+		PullRequests struct {
+			PageInfo struct {
+				HasNextPage githubql.Boolean
+				EndCursor   githubql.String
+			}
+			Nodes []struct {
+				PullRequest PullRequest `graphql:"... on PullRequest"`
+			}
+		} `graphql:"pullRequests(first: 100, after: $prsCursor, states: $prsStates)"`
+	}
+}
+
+// Returns new user dashboard agent.
+func NewDashboardAgent(config *config.GithubOAuthConfig, log *logrus.Entry) *DashboardAgent {
+	return &DashboardAgent{
+		goac: config,
+		log:  log,
+	}
+}
+
+// HandleUserDashboard returns a http handler function that handles request to /user-dashboard
+// endpoint. The handler takes user access token stored in the cookie to query to Github on behalf
+// of the user and serve the data in return. The Query handler is passed to the method so as it
+// can be mocked in the unit test..
+func (da *DashboardAgent) HandleUserDashboard(queryHandler PullRequestQueryHandler) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		serverError := func(action string, err error) {
+			da.log.WithError(err).Errorf("Error %s.", action)
+			msg := fmt.Sprintf("500 Internal server error %s: %v", action, err)
+			http.Error(w, msg, http.StatusInternalServerError)
+		}
+
+		session, err := da.goac.CookieStore.Get(r, da.goac.GithubTokenSession)
+		if err != nil {
+			serverError("Error with getting git token session.", err)
+			return
+		}
+		token, ok := session.Values[da.goac.GithubTokenKey].(*oauth2.Token)
+		data := UserData{
+			Login: false,
+		}
+		if ok && token.Valid() {
+			data.Login = true
+			ghc := github.NewClient(token.AccessToken, githubEndpoint)
+			pullRequests, err := queryHandler.Query(context.Background(), ghc)
+			if err != nil {
+				serverError("Error with querying user data.", err)
+				return
+			} else {
+				data.PullRequests = pullRequests
+			}
+		}
+
+		marshaledData, err := json.Marshal(data)
+		if err != nil {
+			da.log.WithError(err).Error("Error with marshalling user data.")
+		}
+
+		w.Write(marshaledData)
+	}
+}
+
+// Query function that returns a list of open pull requests owned by the user whose access token
+// is consumed by the github client.
+func (da *DashboardAgent) Query(ctx context.Context, ghc githubClient) ([]PullRequest, error) {
+	var prs = []PullRequest{}
+	vars := map[string]interface{}{
+		"prsStates": []githubql.PullRequestState{githubql.PullRequestStateOpen},
```

Comment:
Does this serialize properly?
Also, I don't think this needs to be parameterized at all. We will only ever care about open PRs so I would hard code that into the query instead of making it a variable.

Reply:
done